### ISSUE TITLE
Improve performance while hashing files during tracing

### ISSUE THREAD

**Issue by mikhailswift** [in-toto/witness#83]
#75 introduced file hashing of opened files by tracees.  This allows us to tell exactly what went into a build, but it comes at a cost.  Calculating hashes is not cheap.  This flame graph of a cpu profile of tracing a linux kernel build shows we're eating up most of our time while calculating digest sets for files opened by tracees.  

![image](https://user-images.githubusercontent.com/3218582/147313881-612a76c9-9793-4020-bca7-b44fc84c6da8.png)

Attached are two cpuprofiles generated from witness running with and without hashing files during tracing, both witnessing a linux kernel build.  To use these, run `go tool pprof -http=":8000" ./witness ./cpuprofile-with-hashing`  

The witness binary used for cpuprofile-with-hashing was built with commit a28645c.  The witness binary used for cpuprofile-without-hashing was built with commit 76bcdaa.  Both builds had commit b20275c cherry picked onto them for the cpu profiling.

Some ideas we can possibly use to help performance here and still gather the file hashes:

1. When a tracee opens a file only for read: calculate the hash in a separate go routine and allow the tracee to continue.
2. For files that are opened multiple times: observe their last modified time to see if we should recalculate the digest
3. For files created/written by tracees: don't record their digests until close, and do so on a separate go routine.

There may be other areas for increases and not all of these may be feasible.

[cpuprofiles.tar.gz](https://github.com/testifysec/witness/files/7772495/cpuprofiles.tar.gz)

**Comment by colek42** [comment]
I don't think we should calculate the hashes on a separate goroutine unless we do something fancy like copy the file before hashing.

Where performance is important, we should direct users to the BPF implementation

**Comment by mikhailswift** [comment]
Do you have a reasoning for why you think we shouldn't?
